File: src/main/java/org/zenbot/szolnok/timetable/batch/batch/step/bus/processor/JsoupDocumentToTimetableProcessor.java
package org.zenbot.szolnok.timetable.batch.batch.step.bus.processor;

import lombok.extern.slf4j.Slf4j;
import org.jsoup.nodes.Document;
import org.springframework.batch.item.ItemProcessor;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.stereotype.Component;
import org.zenbot.szolnok.timetable.batch.configuration.properties.TimetableProperties;
import org.zenbot.szolnok.timetable.batch.configuration.properties.TimetableSelectorProperties;
import org.zenbot.szolnok.timetable.batch.domain.Timetable;

import java.util.Map;

@Slf4j
@Component
@EnableConfigurationProperties(TimetableProperties.class)
public class JsoupDocumentToTimetableProcessor implements ItemProcessor<Document, Timetable> {

    public static final String WEEKDAY_KEY = "weekday";
    public static final String SATURDAY_KEY = "saturday";
    public static final String SUNDAY_KEY = "sunday";

    private final TimetableSelectorProperties selectorProperties;

    private final DocumentSelector documentSelector;
    private final EndBusStopSelectorItemProcessorHelper endBusStopSelectorItemProcessorHelper;
    private final ActualStopSelectorItemProcessorHelper actualStopSelectorItemProcessorHelper;
    private final TimetableRowBuilderItemProcessorHelper timetableRowBuilderItemProcessorHelper;

    public JsoupDocumentToTimetableProcessor(TimetableProperties properties, DocumentSelector documentSelector, EndBusStopSelectorItemProcessorHelper endBusStopSelectorItemProcessorHelper, ActualStopSelectorItemProcessorHelper actualStopSelectorItemProcessorHelper, TimetableRowBuilderItemProcessorHelper timetableRowBuilderItemProcessorHelper) {
        this.selectorProperties = properties.getSelector();
        this.documentSelector = documentSelector;
        this.endBusStopSelectorItemProcessorHelper = endBusStopSelectorItemProcessorHelper;
        this.actualStopSelectorItemProcessorHelper = actualStopSelectorItemProcessorHelper;
        this.timetableRowBuilderItemProcessorHelper = timetableRowBuilderItemProcessorHelper;
    }

    @Override
    public Timetable process(Document htmlDocument) {
        Timetable timetable = new Timetable();
        String busName = documentSelector.getText(htmlDocument, selectorProperties.getRouteNameSelector());
        String startBusStop = documentSelector.getText(htmlDocument, selectorProperties.getFromSelector());
        String endBusStop = endBusStopSelectorItemProcessorHelper.getEndBusStop(htmlDocument, selectorProperties);
        String actualStop = actualStopSelectorItemProcessorHelper.getActualStop(htmlDocument, selectorProperties.getActualStopSelector());
        Map<Integer, Map<String, String>> timetableRows = timetableRowBuilderItemProcessorHelper.getTimetableRows(htmlDocument, selectorProperties);

        timetable.setBusName(busName);
        timetable.setStartBusStopName(startBusStop);
        timetable.setEndBusStopName(endBusStop);
        timetable.setActiveStopName(actualStop);
        timetableRows.forEach(timetable::addRow);

        log.info("Process html with busName [#{}] and timetable for [{}] stop", timetable.getBusName(), timetable.getActiveStopName());
        return timetable;
    }

}

File: szolnok-timetable-batch-bus/src/main/kotlin/org/zenbot/szolnok/timetable/batch/bus/batch/step/bus/processor/JsoupDocumentToTimetableProcessor.kt
package org.zenbot.szolnok.timetable.batch.bus.batch.step.bus.processor

import org.jsoup.nodes.Document
import org.slf4j.LoggerFactory
import org.springframework.batch.item.ItemProcessor
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.stereotype.Component
import org.zenbot.szolnok.timetable.batch.utils.configuration.properties.TimetableProperties
import org.zenbot.szolnok.timetable.batch.utils.domain.Timetable

@Component
@EnableConfigurationProperties(TimetableProperties::class)
class JsoupDocumentToTimetableProcessor(
    private val properties: TimetableProperties,
    private val startBusStopSelectorItemProcessorHelper: StartBusStopSelectorItemProcessorHelper,
    private val endBusStopSelectorItemProcessorHelper: EndBusStopSelectorItemProcessorHelper,
    private val actualStopSelectorItemProcessorHelper: ActualStopSelectorItemProcessorHelper,
    private val timetableRowBuilderItemProcessorHelper: TimetableRowBuilderItemProcessorHelper
) : ItemProcessor<Document, Timetable> {

    private val log = LoggerFactory.getLogger(JsoupDocumentToTimetableProcessor::class.java)

    override fun process(htmlDocument: Document): Timetable {
        val timetable = Timetable()
        val busName = htmlDocument.select(properties.selector.routeNameSelector).text()
        val startBusStop = startBusStopSelectorItemProcessorHelper.getStartBusStop(htmlDocument, properties.selector)
        val endBusStop = endBusStopSelectorItemProcessorHelper.getEndBusStop(htmlDocument, properties.selector)
        val actualStop = actualStopSelectorItemProcessorHelper.getActualStop(htmlDocument, properties.selector)
        val timetableRows = timetableRowBuilderItemProcessorHelper.getTimetableRows(htmlDocument, properties.selector)

        timetable.busName = busName
        timetable.startBusStopName = startBusStop
        timetable.endBusStopName = endBusStop
        timetable.activeStopName = actualStop
        timetableRows.forEach { hour, values -> timetable.addRow(hour, values) }

        log.info("Process html with busName [#{}] and timetable for [{}] stop", timetable.busName, timetable.activeStopName)
        return timetable
    }

    companion object {

        val WEEKDAY_KEY = "weekday"
        val SATURDAY_KEY = "saturday"
        val SUNDAY_KEY = "sunday"
    }
}
